src/lib/updating-element.ts

```diff
@@ -0,0 +1,552 @@
+
+const microtaskPromise = new Promise((resolve) => resolve(true));
+
+const STATE_HAS_UPDATED = 1;
+const STATE_IS_UPDATING = 1 << 2;
+const STATE_IS_REFLECTING = 1 << 3;
+type ValidationState = typeof STATE_HAS_UPDATED | typeof STATE_IS_UPDATING | typeof STATE_IS_REFLECTING;
+
+/**
+ * Base element class which manages element properties and attributes. When
+ * properties change, the `update` method is asynchronously called. This method
+ * should be supplied by subclassers to render updates as desired.
+ */
+export abstract class UpdatingElement extends HTMLElement {
+
+  /**
+   * Maps attribute names to properties; for example `foobar` attribute
+   * to `fooBar` property.
+   */
+  private static _attributeToPropertyMap: AttributeMap = new Map();
+
+  /**
+   * Marks class as having finished creating properties.
+   */
+  private static _finalized = true;
+
+  /**
+   * Memoized result of computed observedAttributes.
+   */
+  private static _observedAttributes: string[]|undefined;
+
+  /**
+   * Memoized list of all class properties, including any superclass properties.
+   */
+  private static _classProperties: PropertyDeclarationMap = new Map();
+
+  static properties: PropertyDeclarations = {};
+
+  /**
+   * Returns a list of attributes corresponding to the registered properties.
+   */
+  static get observedAttributes() {
+    if (!this.hasOwnProperty('_observedAttributes')) {
+      // note: piggy backing on this to ensure we're _finalized.
+      this._finalize();
+      this._observedAttributes = [];
+      for (const [p, v] of this._classProperties) {
+        const attr = this._attributeNameForProperty(p, v);
+        if (attr !== undefined) {
+          this._attributeToPropertyMap.set(attr, p);
+          this._observedAttributes.push(attr);
+        }
+      }
+    }
+    return this._observedAttributes;
+  }
+
+  /**
+   * Creates a property accessor on the element prototype if one does not exist.
+   * The property setter calls the property's `shouldInvalidate` property option
+   * or uses a strict identity check to determine if the set should trigger
+   * invalidation and update.
+   */
+  static createProperty(name: PropertyKey, options: PropertyDeclaration = defaultPropertyDeclaration) {
+    // ensure private storage for property declarations.
+    if (!this.hasOwnProperty('_classProperties')) {
+      this._classProperties = new Map();
+      // NOTE: Workaround IE11 not supporting Map constructor argument.
+      const superProperties = Object.getPrototypeOf(this)._classProperties;
+      if (superProperties !== undefined) {
+        superProperties.forEach((v: any, k: PropertyKey) =>
+          this._classProperties.set(k, v));
+      }
+    }
+    this._classProperties.set(name, options);
+    // Allow user defined accessors by not replacing an existing own-property accessor.
+    if (this.prototype.hasOwnProperty(name)) {
+      return;
+    }
+    const key = typeof name === 'symbol' ? Symbol() : `__${name}`;
```

Comment:
Seems like the key should be a Symbol either way

Reply:
Will perf test.